core/jvm/src/main/scala/zio/zmx/internal/ConcurrentSummary.scala

```diff
@@ -31,9 +29,10 @@ object ConcurrentSummary {
 
   def manual(maxSize: Int, maxAge: Duration, error: Double, quantiles: Chunk[Double]): ConcurrentSummary =
     new ConcurrentSummary {
-      private val values          = new ConcurrentLinkedDeque[(java.time.Instant, Double)]
+      private val values = new Array[(java.time.Instant, Double)](maxSize)
```

Comment:
Think you want `AtomicReferenceArray` since each element needs to be updated atomically.

Reply:
It is used as a ring buffer and each element is only written once, and read during snapshot. I think using a simple array for this is safe.

Reply:
I think this is still a risk that we observe inconsistent data here. If we are reading from an index and someone is concurrently writing to that index at the same we could observe a partial update since there is nothing guaranteeing that the update is performed atomically.

Also, I don't think there is anything right now creating a "happens after" relationship between observing a metric and getting a snapshot. We write to `head` in `observe` but we do that before updating the `Array` and then we don't read from `head` at all in `snapshot`.

Reply:
How does such a partial update state look like? I cannot imagine it :) The array cell is a reference and it will either point to a past tuple (or null) or the new one.

I know this implementation looks too simple and I was surprised by it too but I believe it is correct. Let me write a few examples:

### When buffer is not full yet
When `head < maxSize` then the beginning of the array contains `head` references to `(timestamp, value)` pairs and the rest of it is `null`. 
If `snaphot` is called it enumerates all the items that are not null and within the time window.
If during that `observe` is called, if the head increment _and_ the array update were already done then it's the same as above. 
If `snapshot` is happening when head is already increased but the array is not updated yet then the new cell pointed by `head` is still `null` and the snapshot will be correct.

The only thing not guaranteed is that `sum` and `count` and the snapshot calculations are not atomic but that I believe was not the goal (and I don't see how it could be done without some kind of lock).

### When the buffer is full
When `head >= maxSize` then the buffer is full of old values and we are overwriting old values. 
In stable state each array cell holds a reference to a valid `(timestamp, value)` pair and the `head+1` cell holds the one with the oldest timestamp (assuming that the timestamps are continuously increasing between each `observe`). 

In the stable case getting all the elements from the array simply means all the known values which we have `maxSize`
of. Their order does not matter because there is a sort-by-timestamp step later. So we take all the known values, filter them by the time window and use the available items to calculate the snapshot.

If `snapshot` is called between the head increment and the array update then it is simply a snapshot of the previous state as `head` is not even used by the snapshot algorithm.

Reply:
Yes you are right about the updates since references are updated atomically. However, the fact that we never create a "happens after" relationship means there is no guarantee that a thread will ever see an update that has been made by another thread.

I think there is also a case where the first thread calls `getAndIncrement` on `head`, then gets rescheduled, then a bunch of other updates occur and we wrap around, another thread updates the same index, and the original thread then overwrites that update with an older one.

Reply:
Yes you are right about the second one - if the buffer size is not large enough and there are many concurrent updates then it is possible that an older observed value overwrites a newer one. I think this is an acceptable tradeoff to avoid locking. Compared to the old (not working) implementation the difference is that there is a chance that not the oldest element gets "removed" but the quantiles are still calculated from a proper time window.

I don't understand your first point yet.

Reply:
As far as I can see, there is no update to each element of the array: only a "set" which clobbers the old value. Therefore, in the worst case, we get transient inconsistency in a snapshot. I doubt that is fatal for a metrics application.

Reply:
Agree with your point about the metrics potentially only needing to be approximately accurate.

Reply:
I think we still have the issue about not doing anything to publish the changes. We could have Thread A call `observe` a bunch of times and Thread B call `snapshot` and Thread A could still observe a state of the `Array` that did not reflect those updates. Arguably the metrics don't have to be completely accurate but I think right now the state that B sees could be arbitrarily out of date.

The issue with overwriting is debatable I think. The metrics potentially only need to be approximately accurate but what if the metric we overwrote turned out to be particularly important (e.g. a single response that took a long time could dramatically change the 99th percentile for service times).

Reply:
When you say thread B will still not see the changes, you mean because of some kind of cpu level caching? 

We could store snapshots after every n'th `update` at the end of update, and poll that but that would mean the polling interval has no effect on the metrics and also it would be one more parameter to the summary to be specified. I personally don't think it is a good idea. Except if it is updated after _each_ `update`, or in other words if we can continuously calculate the full snapshot. It feels to heavy for me. 

If we want to guarantee that no value get lost then we would have to block when the buffer is full, would not we?

Reply:
For the first problem yes if one thread updates a non-volatile variable there is no guarantee that another thread will ever see that update. We could potentially address just by making the array `volatile`.

For the second, you're right that we could always lose updates if we go over the buffer capacity. But at least ideally the guarantee would be that we would lose the oldest updates, not the most recent updates.

If the buffer has a maximum capacity of 100 and 101 events ago there was some really bad service time then dropping that event and showing a "good" 99th percentile service time for the window is the right answer. But if the most recent event had a really bad service time then the right answer is to reflect that in the metrics.

Reply:
I agree that would be a better behavior and would fully agree if this implementation would always (or very frequently) drop the recent values.

I think that is a rare (but possible) case that happens if the updates are very frequent compared to the snapshot and it depends on the buffer size - by increasing the buffer the probability can be decreased. So I am not convinced that this is an issue justifying a more complex / slow implementation. (At least I could not come up with any good solutions yet ;)

Reply:
Seems like making the array volatile does not guarantee anything for the item updates so I replaced it with `AtomicReferenceArray`

Reply:
We could just "publish" our updates by writing to some volatile variable after writing a metric and read from that same variable before taking a snapshot.

I think you're doing exactly the right thing in pushing for performance and where we can relax guarantees to get it I just want to make sure we understand the trade-offs involved and are getting to the right place there.

Reply:
Late to the game - just ha a quick look at the changes and the tests. LGTM. The general idea for a summary according to the Prometheus spec (at least from my understanding) is that it happens to reflect the values that are currently in the buffer - even if that means we would be dropping elements that have never been used fro reporting. At the end it comes down to optimising the buffer size and the reporting interval. 

Even with "Push"-like protocols we might not see certain data. Here the data is pushed periodically to an agent, but that agent accumulates data in intervals as well - and might be further aggregated within the dashboard. 

Therefore I think using an `AtomicReferenceArray` with a fixed length as suggested will work well for our Summary.